ci: Fix COMMIT_RANGE variable value for PRs

hebasto opened the pull request:
This PR:
- is a #20658 and #20682  followup 
- set the `COMMIT_RANGE` variable correctly for PRs
- cleans up Travis-specific code
- prints COMMIT_RANGE value to the log for convenience as it was in Travis CI

hebasto committed:
ci: Fix COMMIT_RANGE variable value for PRs

hebasto committed:
ci: Drop Travis-specific way to set COMMIT_RANGE variable

hebasto committed:
ci: Drop Travis-specific workaround for shellcheck

MarcoFalke commented on ci/lint/06_script.sh:
```diff
@@ -7,12 +7,12 @@
+  COMMIT_RANGE="$CIRRUS_BASE_BRANCH..HEAD"
   test/lint/commit-script-check.sh $COMMIT_RANGE
+else
+  COMMIT_RANGE="$(git fetch $CIRRUS_REPO_CLONE_URL $CIRRUS_LAST_GREEN_CHANGE && git merge-base $CIRRUS_LAST_GREEN_CHANGE HEAD)..HEAD"
```
we don't run the linters on any branch because it is too late to fixup any issues that have been merged

hebasto replied:
Correct. But this is for forked repos and personal Cirrus accounts.

MarcoFalke replied:
Maybe submit this as a separate pull? ;)

hebasto replied:
Thanks! [Done](https://github.com/bitcoin/bitcoin/pull/20697#issuecomment-748600578).

maflcko reviewed (commented):
this still "fails" because `master` doesn't exist when the repo was cloned with `go`. I think creating the master branch after `git fetch` will do.

MarcoFalke commented on ci/lint/05_before_script.sh:
```diff
@@ -6,4 +6,4 @@
 export LC_ALL=C
 
-git fetch
+git fetch $CIRRUS_REPO_CLONE_URL $CIRRUS_BASE_BRANCH
```
The command to create a branch would be git branch master origin/master

MarcoFalke commented on ci/lint/05_before_script.sh:
```diff
@@ -6,4 +6,5 @@
 
-git fetch
+git fetch origin
+git branch $CIRRUS_BASE_BRANCH origin/$CIRRUS_BASE_BRANCH
```
needs to be guarded by ` if [ -n "$CIRRUS_PR" ]; then`

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/20697#issuecomment-748600578).

hebasto committed:
ci: Print COMMIT_RANGE to the log as it was in Travis CI

hebasto commented:
Updated 144250e3f055d47d18379b01056e9d3eb11e9231 -> 83c07d1e3f8809183cce4ca88176d700249e9f55 ([pr20697.03](https://github.com/hebasto/bitcoin/commits/pr20697.03) -> [pr20697.04](https://github.com/hebasto/bitcoin/commits/pr20697.04)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/20697#discussion_r545826069):
> Maybe submit this as a separate pull? ;)

See: #20728.

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/20697#discussion_r545828178):
> needs to be guarded by ` if [ -n "$CIRRUS_PR" ]; then`

- added printing of the `COMMIT_RANGE` variable value to the log as it was in Travis CI

MarcoFalke commented on ci/lint/06_script.sh:
```diff
@@ -6,13 +6,13 @@
-  # by a pull request this is the name of the branch targeted by the pull request.
-  # https://cirrus-ci.org/guide/writing-tasks/#environment-variables
-  COMMIT_RANGE="$CIRRUS_BRANCH..HEAD"
+  git fetch $CIRRUS_REPO_CLONE_URL $CIRRUS_BASE_SHA
```
Is this needed?

hebasto replied:
Actually, it replaces the `git fetch` command from the removed `ci/lint/05_before_script.sh`

Oh, I see. The `CIRRUS_BASE_SHA` must already be in the commit history for pulls, right?
I'll push another change without this command.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/20697#issuecomment-748869426).

maflcko reviewed (approved):
ACK

hebasto commented:
Updated 83c07d1e3f8809183cce4ca88176d700249e9f55 -> b4cd93d6f3e822365b107ccd5f0426fef31b7156 ([pr20697.04](https://github.com/hebasto/bitcoin/commits/pr20697.04) -> [pr20697.05](https://github.com/hebasto/bitcoin/commits/pr20697.05), [diff](https://github.com/hebasto/bitcoin/compare/pr20697.04..pr20697.05)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/20697#discussion_r546541495)

- fixed `COMMIT_RANGE` value when a changed PR is force pushed into the base branch that has already grown, i.e., the pull is effectively based on the non-top commit of the base branch

- added `git log` command to the lint task log

MarcoFalke commented on ci/lint/06_script.sh:
```diff
@@ -6,13 +6,12 @@
-  # by a pull request this is the name of the branch targeted by the pull request.
-  # https://cirrus-ci.org/guide/writing-tasks/#environment-variables
-  COMMIT_RANGE="$CIRRUS_BRANCH..HEAD"
+  COMMIT_RANGE="$(git merge-base $CIRRUS_BASE_SHA $CIRRUS_CHANGE_IN_REPO)..$GIT_HEAD"
```
Why is this changed again? Now it will include master commits, because master has already been merged in for pull requests (see .cirrus.yml)

`COMMIT_RANGE="${CIRRUS_BASE_SHA}..$GIT_HEAD"` was correct.

hebasto replied:
> ... master has already been merged in for pull requests (see .cirrus.yml)

Right.

For described case "when a changed PR is force pushed into the base branch that has already grown, i.e., the pull is effectively based on the non-top commit of the base branch" I think `COMMIT_RANGE="${CIRRUS_BASE_SHA}..$GIT_HEAD"` is wrong, as `$CIRRUS_BASE_SHA` is not the commit on which the pull is based on.

MarcoFalke replied:
> $CIRRUS_BASE_SHA is not the commit on which the pull is based on.

It is the commit that this pull is de-facto based on, assuming that `git merge A` is approximately doing the same like `git rebase A`.

As I mentioned previously the current version is broken and will include all commits that have been pushed to master since. See:

https://cirrus-ci.com/task/5396234322051072?command=lint#L931

hebasto replied:
In `.cirrus.yml` the base branch is merged into the pull one. That's why I think that [listed](https://cirrus-ci.com/task/5396234322051072?command=lint#L931) commits are correct `COMMIT_RANGE` to lint.

OTOH, we could merge the pull branch into the base one. In that case `$CIRRUS_BASE_SHA` will work as documented.

MarcoFalke replied:
> That's why I think that listed commits are correct COMMIT_RANGE to lint.

It is not the way it worked on travis

hebasto replied:
Agree.

Is it acceptable to s/`git merge A`/`git rebase A`/ ?

hebasto replied:
I've tested `COMMIT_RANGE="${CIRRUS_BASE_SHA}..$GIT_HEAD"`. It works for all cases.
`git` is smart enough to know what to do :)

But for human the value of `COMMIT_RANGE` is a bit cryptic (regarding to pull commits) in cases of a pull based on the non-top commit of the base branch. So going to skip output of it.

MarcoFalke replied:
> Is it acceptable to s/git merge A/git rebase A/ ?

In theory yes, in practise it doesn't work with subtree bumps.

MarcoFalke commented:
ACK 3c2478c38522c176e81befd4d991a259b09be063

hebasto commented:
Updated b4cd93d6f3e822365b107ccd5f0426fef31b7156 -> f89d374069e2c4b7897d4b95f69876fc42604550 ([pr20697.05](https://github.com/hebasto/bitcoin/commits/pr20697.05) -> [pr20697.07](https://github.com/hebasto/bitcoin/commits/pr20697.07), [diff](https://github.com/hebasto/bitcoin/compare/pr20697.05..pr20697.07)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin/bitcoin/pull/20697#discussion_r546611033):
> Why is this changed again? Now it will include master commits, because master has already been merged in for pull requests (see .cirrus.yml)
> 
> `COMMIT_RANGE="${CIRRUS_BASE_SHA}..$GIT_HEAD"` was correct.

MarcoFalke commented:
This was merged two seconds before you force pushed :sweat_smile:

hebasto commented:
> This was merged two seconds before you force pushed sweat_smile

slow typing :)
